store/storage/db.go

```diff
@@ -0,0 +1,221 @@
+package storage
+
+import (
+	"fmt"
+	"sync"
+
+	"golang.org/x/exp/maps"
+	"golang.org/x/exp/slices"
+
+	"cosmossdk.io/store/v2"
+)
+
+// TODO:
+//
+// - Define and implement pruning method(s)
+// - Import/snapshot method(s)???
+// - Should we make this an interface?
+//
+// Ref: https://github.com/cosmos/cosmos-sdk/issues/17279
+
+// DefaultMemDBSize defines the default pre-allocation size of the in-memory
+// database used to accumulate database entries prior to committing them to disk.
+const DefaultMemDBSize = 1000
+
+var (
+	_ store.VersionedReaderWriter    = (*Database)(nil)
+	_ store.VersionedIteratorCreator = (*Database)(nil)
+	_ store.Committer                = (*Database)(nil)
+)
+
+// Database defines the state storage (SS) backend. Each open instance of a Database
+// is meant to commit key/value pairs to disk upon Commit for a single version.
+// However, it can be used to read key/value pairs at any version.
+type Database struct {
+	mu      sync.RWMutex
+	vdb     store.VersionedDatabase
+	version uint64
+	memSize int
+	mem     map[string]map[string]dbEntry
+	batch   store.Batch
+}
+
+// dbEntry defines a database entry stored in memory until the batch is written
+// on commit.
+type dbEntry struct {
+	value   []byte
+	delete  bool
+	version uint64
+}
+
+// New creates a new instance of a state storage (SS) backend Database and returns
+// a reference to it. It takes a backing physical storage backend, the current
+// version/height that will be committed, the list of store keys to support and
+// the pre-allocation size for each store key.
+func New(vdb store.VersionedDatabase, version uint64, storeKeys []string, memSize int) *Database {
+	db := &Database{
+		vdb:     vdb,
+		version: version,
+		memSize: memSize,
+		batch:   vdb.NewBatch(version),
+	}
+
+	for _, storeKey := range storeKeys {
+		db.mem[storeKey] = make(map[string]dbEntry, memSize)
+	}
+
+	return db
+}
+
+func (db *Database) Close() error {
+	db.mu.Lock()
+	defer db.mu.Unlock()
+
+	if err := db.vdb.Close(); err != nil {
+		return err
+	}
+
+	// reset all mutable fields
+	db.vdb = nil
+	db.mem = nil
+	db.batch = nil
+
+	return nil
+}
+
+func (db *Database) Has(storeKey string, version uint64, key []byte) (bool, error) {
+	db.mu.RLock()
+	defer db.mu.RUnlock()
+
+	if db.mem == nil {
+		return false, store.ErrClosed
+	}
+	// TODO: Should we enforce reads have a store key set?
+
+	if entry, ok := db.mem[storeKey][string(key)]; ok && entry.version == version {
+		return !entry.delete, nil
+	}
+
+	return db.vdb.Has(storeKey, version, key)
+}
+
+func (db *Database) Get(storeKey string, version uint64, key []byte) ([]byte, error) {
+	db.mu.RLock()
+	defer db.mu.RUnlock()
```

Comment:
We discourage using sync.RWMutex as you really can't predict right now the ratio of reads to writes and sync.RWMutex scales poorly. Please take a look at our recommendation https://cyber.orijtech.com/scsec/cosmos-go-coding-guide#avoid-syncrwmutex